### PR TITLE
AG-16878 reduce formula data service calls

### PR DESCRIPTION
**Pull request by SalvatorePreviti**
## Summary

Reduces redundant calls to `formulaDataSource.getFormula` and `valueService.getValue` from the formula engine, narrows `cellValueChanged` invalidation from a full cache wipe to a row-local eviction, and adds a grid API to invalidate the formula cache explicitly when the external store is mutated out-of-band.

I won't really define the new caching mechanism as a breaking change as I doubt users expected the data source to be automatically reflected without a grid refresh / edit / setData, but, providing the new API gives them an option to do this.
The right behaviour is to cache the data source results, so this is a bug fix more than a breaking change - and providing this new API as an escape hatch if the user would ever need to.

## New API: `api.refreshFormulas`

In-grid edits and CSRM updates already auto-invalidate the formula cache. Users who mutate the external formula store out-of-band (e.g. backend sync) previously had no way to tell the grid to re-query. New grid API:

- `api.refreshFormulas()` — invalidates every cached formula.
- `api.refreshFormulas(rowNode)` — invalidates a single row (plus pinned / group-footer siblings) given an `IRowNode`.
- `api.refreshFormulas(rowId)` — same, but resolves the row via `rowModel.getRowNode(id)`; unknown id is a silent no-op.

Documented under the Formula Data Source / Caching section in the Formulas docs.

## Changes

- **Single source of truth for per-cell lookups.** `ensureCellFormula` is now the only place that queries `dataSource.getFormula` or reads the raw cell value. `valueService` (for the column-value path), `getDataSourceFormula`, `getFormulaError`, `resolveValue`, and the eval loop all route through it.
- Changes the lookup in value service for non formulas and formulas, the new implementation make the performance critical getValue a bit faster, as benchmarks shows too.
- **Negative caching.** Plain-value cells in `allowFormula` columns cache a `null` result, so N dependent formulas referencing the same plain cell trigger one lookup, not N.
- **Re-entry guard.** `ensureCellFormula` writes a `null` sentinel before calling out, so a `valueGetter` that synchronously calls back via `api.getCellValue` for the same cell sees "not a formula" instead of recursing. On throw the sentinel is cleared so the next read retries.
- **`fromDataSource` flag on `CellFormula`.** Lets `getDataSourceFormula` distinguish formulas produced by the data source from `"=..."` strings in raw row data without re-querying the data source.
- **Row-local cache eviction on `cellValueChanged`.** Drops the edited row's map and bumps the value-version counter instead of wiping the whole cache. Relies on `valueService.finishValueChange` already firing the event for both the row and its pinned sibling.
- **Resolver refactor.** `FormulaService` implements a new `FormulaResolver` interface (`ensureCellFormula` + `resolveAddrRef`) so `evalAst` / `unresolvedDeps` / range iterators take the service directly, eliminating per-frame bound-callback allocations and the `activeCtx` arrow-field closures. Cycle-detection state moves to a plain `Map<RowNode, Set<AgColumn>>`.
- **Error propagation.** `resolveAddrRef` now throws the `CellFormula` itself for errored dependencies; the outer catch decomposes it, avoiding a `FormulaError` allocation on the eval hot path.
- **Shared helpers.** A1-column alphabet is hoisted to a module-level constant; bigint coercion is delegated to the shared `_parseBigIntOrNull`.
- Improved documentation grammar and wording and avoid "you".

## Behaviour changes

- `cellValueChanged` no longer wipes the entire formula cache; only the edited row's entry (and its pinned sibling, via the separate event `valueService` already dispatches) is dropped. Cross-row `getFormula` dependencies that relied on the old full-wipe behaviour will see stale formula text — these now need `api.refreshFormulas()`.
- `getFormulaError` no longer calls `resolveValue` for cells that aren't formulas;

## Tests

Eight new behavioural tests in `formulas-edge-cases.test.ts`:

- `getFormula` is only called for `allowFormula` columns.
- Formulas can `REF` into non-`allowFormula` columns without triggering `getFormula` there.
- Plain-value cells cache their "not a formula" result (one `getFormula` call shared across N references).
- `"=..."` raw values on non-`allowFormula` columns are not parsed as formulas (also via `REF`).
- `api.refreshFormulas()` invalidates the whole cache and re-queries every cell.
- `api.refreshFormulas(rowNode)` invalidates only that row.
- `api.refreshFormulas(rowId)` resolves string ids via the row model.
- `api.refreshFormulas(unknownId)` is a silent no-op; cached values stay as-is.

## Benchmarks 

compared to the version before the prev AG-16878 PR, so, before the work on this ticket started:

↑ 3.89x faster [formulas] cold: invalidate then evaluate sum/product/branch for 3000 rows (16.48 → 64.08 ops/s)
↑ 2.40x faster [formulas] update source -> re-eval 200 chained + 1500 fan-out dependents (148.0 → 355.3 ops/s)
↑ 1.30x faster [grouping-aggregation] immutable (5/50 cols) — CellsPath (75.23 → 97.44 ops/s)
↑ 1.28x faster [grouping-aggregation] transaction (5/50 cols) — RowsPath (79.42 → 101.3 ops/s)
↑ 1.27x faster [grouping-aggregation] immutable (all 50 cols) — RowsPath (72.65 → 92.44 ops/s)
↑ 1.27x faster [grouping-aggregation] transaction (5/50 cols) — CellsPath (79.27 → 100.5 ops/s)
↑ 1.22x faster [formulas] cold: invalidate then SUM over 100000 cells (48.97 → 59.92 ops/s)
↑ 1.21x faster [flat-sorting] sort 30000 rows (26.13 → 31.66 ops/s)
↑ 1.12x faster [tree-data-path] build from scratch 20884 rows (7.381 → 8.244 ops/s)
↑ 1.10x faster [delta-sort-transaction] applyTransaction (deltaSort: false) - 1 update (7.194 → 7.911 ops/s)
↑ 1.09x faster [getvalue] getCellValue by string (last of 100 cols) (9373.8 → 10,256 ops/s)
↑ 1.09x faster [flat-pipeline] filter + sort 30000 rows (18.13 → 19.70 ops/s)
↑ 1.08x faster [getvalue] getDataValue by Column object (first col) (12,630 → 13,608 ops/s)
↑ 1.06x faster [delta-sort-transaction] applyTransaction (deltaSort: true) - 1 update (60.78 → 64.69 ops/s)

FIX AG-16878

### CODE REVIEW

File: packages/ag-grid-enterprise/src/formula/formulaService.ts [hunk 1]
```diff
@@ -183,10 +161,17 @@ export class FormulaService extends BeanStub implements IFormulaService, NamedBe
...
+            if (!this.active) {
+                return;
             }
+            // Drop the whole row's map, not just the edited cell: `getFormula` may derive
+            // cell B's formula string from other cells' data in the same row, so editing A
+            // can change B's text — and `CellFormula.formulaString` never refreshes without eviction.
+            // Pinned counterpart receives its own cellValueChanged event; group-footer
+            // siblings never cache (formulas skipped for group rows).
+            this.cachedResult.delete(event.node as RowNode);
+            this.bumpValueCacheAndRefresh();
```
[the remark this answers is not in the document]
Comment (SalvatorePreviti) [reply packages/ag-grid-enterprise/src/formula/formulaService.ts:183]: this is by design and the scope of this pull request - invalidate only the edited row and not the whole grid

File: documentation/ag-grid-docs/astro.config.mjs [hunk 2]
```diff
@@ -161,21 +161,6 @@ const httpsEnabled = !['0', 'false'].includes(PUBLIC_HTTPS_SERVER);
...
-         * have to be allowed here as well as in `vite.server.cors.origin` below.
-         */
-        allowedDomains: [
-            // Plunkr
-            { hostname: 'run.plnkr.co', protocol: 'https' },
-            // Codesandbox
-            { hostname: '**.csb.app', protocol: 'https' },
-        ],
-    },
     devToolbar: {
```
[the remark this answers is not in the document]
Comment (SalvatorePreviti) [reply documentation/ag-grid-docs/astro.config.mjs:179]: merging issue resolved

File: packages/ag-grid-enterprise/src/formula/formulaService.ts [hunk 3]
```diff
@@ -183,10 +165,22 @@ export class FormulaService extends BeanStub implements IFormulaService, NamedBe
...
             }
+            // Always bump: a formula in another row may REF a non-formula column here, whose
+            // cells never enter this cache — only the version bump invalidates those computed
+            // values. `dropRow` additionally evicts this row's and its pinned sibling's own
+            // CellFormula entries so their captured `formulaString` re-queries `getFormula`.
+            this.dropRow(node);
+            this.bumpValueCacheAndRefresh();
         };
         const onNewColumnsLoaded = () => {
             if (!this.active) {
```
[the remark this answers is not in the document]
Comment (SalvatorePreviti) [reply packages/ag-grid-enterprise/src/formula/formulaService.ts:186]: this is by design - and the point of the whole PR. Documentation and test added

**Review by StephenCooper — APPROVED** [review]
